filebeat/harvester/readers.go

```diff
@@ -0,0 +1,242 @@
+package harvester
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type lineReader interface {
+	Next() ([]byte, int, error)
+}
+
+// noEOLLineReader strips last EOL from input line reader (if present).
+type noEOLLineReader struct {
+	reader lineReader
+}
+
+type logFileReader struct {
+	fs        FileSource
+	offset    int64
+	config    logFileReaderConfig
+	truncated bool
+
+	lastTimeRead time.Time
+	backoff      time.Duration
+}
+
+type timeoutLineReader struct {
+	reader  lineReader
+	timeout time.Duration
+	signal  error
+
+	running bool
+	ch      chan lineMessage
+}
+
+type lineMessage struct {
+	line []byte
+	sz   int
+	err  error
+}
+
+type logFileReaderConfig struct {
+	forceClose         bool
+	maxInactive        time.Duration
+	backoffDuration    time.Duration
+	maxBackoffDuration time.Duration
+	backoffFactor      int
+}
+
+var (
+	errFileTruncate = errors.New("detected file being truncated")
+	errForceClose   = errors.New("file must be closed")
+	errInactive     = errors.New("file inactive")
+	errTimeout      = errors.New("timeout")
+)
+
+func newNoEOLLineReader(r lineReader) *noEOLLineReader {
+	return &noEOLLineReader{r}
+}
+
+func (n *noEOLLineReader) Next() ([]byte, int, error) {
```

Comment:
Add some docs here, that this line reader removes the newline chars. Perhaps it is better named noNewlineLineReader.